module/mempool/herocache/backdata/heropool/pool_test.go

```diff
@@ -215,6 +231,97 @@ func TestInvalidateEntity(t *testing.T) {
+		ejectionMode      EjectionMode        // ejection mode
+		operationsIndexes []OperationAndIndex // operation to perform on an entity stored under the given index,
+		// where index ranges from 0 to entityCount
+		finalEntitiesInThePool []uint // indexes of entities residing in the pool at the end of the test
+	}{
+		{
+			limit:                  2,
+			entityCount:            5,
+			ejectionMode:           LRUEjection,
+			operationsIndexes:      []OperationAndIndex{{Add, 0}, {Add, 0}, {Add, 1}, {Add, 2}, {Add, 3}, {Remove, 2}, {Add, 2}, {Remove, 2}},
+			finalEntitiesInThePool: []uint{3},
+		},
+		{
+			limit:                  2,
+			entityCount:            5,
+			ejectionMode:           NoEjection,
+			operationsIndexes:      []OperationAndIndex{{Add, 0}, {Add, 1}, {Add, 2}, {Add, 3}, {Remove, 0}, {Remove, 1}, {Add, 4}, {Add, 3}},
+			finalEntitiesInThePool: []uint{3, 4},
+		},
+		{
+			limit:                  5,
+			entityCount:            6,
+			ejectionMode:           RandomEjection,
+			operationsIndexes:      []OperationAndIndex{{Add, 0}, {Add, 1}, {Remove, 1}, {Add, 2}, {Remove, 0}, {Add, 5}, {Add, 4}, {Add, 1}, {Add, 3}, {Add, 0}},
+			finalEntitiesInThePool: nil,
+		},
+	} {
+		t.Run(fmt.Sprintf("%d-limit-%d-entities", tc.limit, tc.entityCount), func(t *testing.T) {
+			testAddRemoveEntities(t, tc.limit, tc.entityCount, tc.ejectionMode, tc.operationsIndexes, tc.finalEntitiesInThePool)
+		})
+	}
+}
+
+// testAddRemoveEntities allows to add or remove entities in an order given by operationsIndexes. Each OperationAndIndex consists of an operaton to perform
+// on an entity stored under a corresponding index. Index range from 0 to entityCount.
+func testAddRemoveEntities(t *testing.T, limit uint32, entityCount uint32, ejectionMode EjectionMode, operationsAndIndexes []OperationAndIndex, finalEntitiesInThePool []uint) {
+
+	pool := NewHeroPool(limit, ejectionMode)
+
+	entities := unittest.EntityListFixture(uint(entityCount))
+
+	// this map maintains entities currently stored in the pool
+	var insertedEntities map[flow.Identifier]EIndex
+	insertedEntities = make(map[flow.Identifier]EIndex)
+
+	for _, operationAndIndex := range operationsAndIndexes {
+		operation := operationAndIndex.operation
+		index := operationAndIndex.index
+		switch operation {
+		case Add:
+			indexInThePool, _, ejectedEntity := pool.Add(entities[index].ID(), entities[index], uint64(index))
+			if indexInThePool != InvalidIndex {
+				insertedEntities[entities[index].ID()] = indexInThePool
+			}
+			if ejectedEntity != nil {
+				delete(insertedEntities, ejectedEntity.ID())
+			}
+		case Remove:
+			indexInPool, found := insertedEntities[entities[index].ID()]
+			if found {
+				removedEntity := pool.Remove(indexInPool)
+				require.Equal(t, entities[index].ID(), removedEntity.ID(), "Removed wrong entity")
+				delete(insertedEntities, entities[index].ID())
+			} else {
+				panic("Ill constructed test, trying to remove an element that has not been inserted")
```

Comment:
```suggestion
				require.Fail("Ill constructed test, trying to remove an element that has not been inserted")
```

Reply:
thank you, will do

Reply:
Please replace all panic statements with `require.Fail`

Reply:
I'll apply your suggestion. Just wanted to let you know that I used panic as this is not a failure of a tested software but reflects an ill formed test.

Reply:
Now,  I realized ,indeed, panic is bad cause may be somewhere on CI it would not allow other tests to be run...